README.md
# chloee
[![Build Status](https://travis-ci.org/heyitsjhu/chloee.svg?branch=master)](https://travis-ci.org/heyitsjhu/chloee)
[![Coverage Status](https://coveralls.io/repos/github/heyitsjhu/chloee/badge.svg?branch=master)](https://coveralls.io/github/heyitsjhu/chloee?branch=master)

A simple hexidecimal and rgb color converter.

## Use Chloee in Node

Install:
```
npm install chloee
```

Then, use it like so:
```
const chloee = require('chloee');

chloee('c470ee');
// '196,112,238'

chloee('196,122,238');
// 'c470ee'
```

## Use Chloee in the Browser
Add the following to your HTML page's `<head>` section.

```
<script src="https://d2uu82licdpgvh.cloudfront.net/chloee.min.js"></script>
```

You can now use Chloee via `chloee()`:
```
var result = chloee('c470ee');
console.log(result);    // '196,112,238'

var result = chloee('196,112,238');
console.log(result);    // 'c470ee'
```

## License
[MIT](https://github.com/heyitsjhu/chloee/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [TODO]
- Allow users to enter HEX values with the leading #-sign (e.g., #c470ee).
- Allow users to enter RGB values with the rgb() format (e.g., rgb(196,112,238)).

## [Unreleased]
- Changed: When returning an RGB value, the result will include the 'rgb' prefix and surrounding parentheses. For example, HEX code #C470EE will return rgb(196,112,238) instead of just 196,112,238.

## 1.0.1 - 2018-xx-xx
### Changed
- Moved away from using browserify for building production files in favor of webpack.

## 1.0.0 - 2018-01-07
### Initial Release
- Added: A demo page that users can use to convert hex/rgb values using Chloee.

## 0.0.1 - 2017-12-31
### Pre-Release
- Feature: Given a HEX code, return the corresponding RGB value. Or, vice versa, given an RGB value, return the corresponding HEX code.

test/chloee.test.js
import { expect } from 'chai';
import Chloee from '../src/lib/Chloee';
import chloee from '../src/index';

const TESTS = {
  case1: { 'hex': 'abcdef', 'rgb': '171,205,239' },
  case2: { 'hex': 'c470ee', 'rgb': '196,112,238' },
  case3: { 'hex': '999999', 'rgb': '153,153,153' },
  case4: { 'hex': 'ebd928', 'rgb': '235,217,40' },
  case5: { 'hex': '10de84', 'rgb': '16,222,132' },
  case6: { 'hex': 'ffffff', 'rgb': '255,255,255' },
  case7: { 'hex': 'af9004', 'rgb': '175,144,4' },
  case8: { 'hex': 'c003d9', 'rgb': '192,3,217' },
  case9: { 'hex': 'ee0066', 'rgb': '238,0,102' },
  case10: { 'hex': '000000', 'rgb': '0,0,0' }
}

describe('new Chloee instance', () => {
  it('should return Missing Argument Error if argument is missing', () => {
    expect(function() {
      new Chloee();
    }).to.throw(TypeError);
  });
  it('should return Type Error if argument is not a string', () => {
    expect(function() {
      new Chloee(999999);
    }).to.throw(TypeError);
  });
  it('should return Format Error if HEX contains invalid letters', () => {
    expect(function() {
      new Chloee('abcghi');
    }).to.throw(Error, /Format Error/i);
  });
  it('should return Format Error if HEX is a negative value', () => {
    expect(function() {
      new Chloee('-777777');
    }).to.throw(Error, /Format Error/i);
  });
  it('should return Format Error if RGB has value greater than 255', () => {
    expect(function() {
      new Chloee('279, 200, 200');
    }).to.throw(Error, /Format Error/i);
  });
  it('should return Format Error if RGB contains letters', () => {
    expect(function() {
      new Chloee('135, abc, 200');
    }).to.throw(Error, /Format Error/i);
  });
  it('should return Format Error if RGB contains negative numbers', () => {
    expect(function() {
      new Chloee('-135, -5, 200');
    }).to.throw(Error, /Format Error/i);
  });
});

describe('#convert', () => {
  for (let test in TESTS) {
    describe(`given a hex value of ${TESTS[test].hex}`, () => {
      const result = chloee(TESTS[test].hex);

      it(`should return ${TESTS[test].rgb}`, () => {
        expect(result).to.equal(TESTS[test].rgb);
      });
    });
  }

  for (let test in TESTS) {
    describe(`given an rgb value of ${TESTS[test].rgb}`, () => {
      const result = chloee(TESTS[test].rgb);

      it(`should return ${TESTS[test].hex}`, () => {
        expect(result).to.equal(TESTS[test].hex);
      });
    });
  }
});

src/lib/chloee.js
class Chloee {
  constructor(color, options = {}) {
    this._color = color;
    this._isRGB = false;
    this._isHEX = false;
    this._code = {
      "a": 10,
      "b": 11,
      "c": 12,
      "d": 13,
      "e": 14,
      "f": 15
    }

    this.validate(this._color);
    this.identifyFormat();
  }

  validate(color) {
    if (color === undefined) {
      throw new Error('Missing Argument! Please provide a valide HEX or RGB value.');
    } else if (typeof color !== 'string') {
      throw new Error('Type Error! Argument must be a string.');
    }
  }

  identifyFormat() {
    const RGBRange = '([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])';
    const HEXFormat = new RegExp(/^#*[0-9abcdef]{6}/, 'i');
    const RGBFormat = new RegExp(`^(${RGBRange}[,\s]+){2}${RGBRange}{1,3}`);

    if (this._color.match(HEXFormat)) {
      this._isHEX = true;
    } else if (this._color.match(RGBFormat)) {
      this._isRGB = true;
    } else {
      throw new Error("Format Error! Argument must be either a HEX code (e.g. #c470ee) or RGB value (e.g. 196,112,238).");
    }
  }

  convert() {
    if (this._isRGB) {
      return this.convertTo('hex', this._color);
    } else if (this._isHEX) {
      return this.convertTo('rgb', this._color);
    }
  }

  convertToLetter(number) {
    for (var key in this._code) {
      if (this._code[key] === number) {
        return key;
      }
    }
  }

  convertTo(type, color) {
    let result = [];

    if (type === 'hex') {
      let converted_rgb = [];
      color.split(',').map((char) => {
        let value = Number(char);
        converted_rgb.push(Math.floor(value / 16));
        converted_rgb.push(value % 16);
      });

      // console.log(converted_rgb);
      for (let i=0; i<converted_rgb.length; i++) {

        converted_rgb[i] > 9
        ? result.push(this.convertToLetter(converted_rgb[i]))
        : result.push(converted_rgb[i]);

      }
      return result.join('');

    } else if (type === 'rgb') {
      let converted_hex = color.split('').map((char) => {
        return char.match(/[abcdef]/i) ? this._code[char] : Number(char);
      });
      // console.log(converted_hex);

      for (let i=0; i<6; i=i+2) {
       result.push((converted_hex[i] * 16) + converted_hex[i+1]);
      }
      return result.join(',');
    }
  }

}

export default Chloee;

src/lib/Chloee.js
class Chloee {

  constructor(color) {
    [this._color, this._isRGB, this._isHEX, this._code] = [color, false, false, {
      "a": 10,
      "b": 11,
      "c": 12,
      "d": 13,
      "e": 14,
      "f": 15
    }];

    this.validateAndIdentify(this._color);
  }

  validateAndIdentify(color) {
    const RGBRange = '([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])';
    const HEXFormat = new RegExp(/^#*[0-9abcdef]{6}/, 'i');
    const RGBFormat = new RegExp(`^(${RGBRange}[,\s]+){2}${RGBRange}{1,3}`);

    if (color === undefined) {
      throw new TypeError('Oops! Looks like you forgot to pass in a HEX or RGB value. Chloee requires a valid argument');
    } 
    else if (typeof color !== 'string') {
      throw new TypeError(`Argument must be a string. You passed in ${color}`);
    } 
    else if (!this._color.match(HEXFormat) && !this._color.match(RGBFormat)) {
      throw new Error(`Format Error! Argument must be a HEX code (e.g. #c470ee) or RGB value (e.g. 196,112,238). You entered ${color}`);
    }

    color.match(HEXFormat) ? this._isHEX = true : this._isRGB = true;
  }

  convert() {
    /* istanbul ignore else */
    if (this._isRGB) {
      return this.convertTo('hex', this._color);
    } else if (this._isHEX) {
      return this.convertTo('rgb', this._color);
    }
  }

  convertToLetter(number) {
    for (var key in this._code) {
      if (this._code[key] === number) {
        return key;
      }
    }
  }

  convertTo(type, color) {
    let result = [];

    /* istanbul ignore else */
    if (type === 'hex') {
      let converted_rgb = [];
      color.split(',').map((char) => {
        let value = Number(char);
        converted_rgb.push(Math.floor(value / 16));
        converted_rgb.push(value % 16);
      });

      // console.log(converted_rgb);
      for (let i=0; i<converted_rgb.length; i++) {

        converted_rgb[i] > 9
        ? result.push(this.convertToLetter(converted_rgb[i]))
        : result.push(converted_rgb[i]);

      }
      return result.join('');

    } else if (type === 'rgb') {
      let converted_hex = color.split('').map((char) => {
        return char.match(/[abcdef]/i) ? this._code[char] : Number(char);
      });
      // console.log(converted_hex);

      for (let i=0; i<6; i=i+2) {
       result.push((converted_hex[i] * 16) + converted_hex[i+1]);
      }
      return result.join(',');
    }
  }

}

module.exports = Chloee;
